### ISSUE TITLE
Significant performance loss caused by matrix indexing in jitted function with nested jax operations

### ISSUE THREAD

**Issue by yannra** [google/jax#9366]
Hi all,
I've recently come across a very strange behaviour within a jitted jax function (so far only tested on CPU). The key issue is that the code is becoming insanely slow when a particular array access operation is performed inside a function body but runs a lot faster in other instances which should have comparable computational time. Unfortunately the issue seems to appear in a very specific mix of vmaps, while loops and conds, so that the following minimal example is a couple of lines long.
What this example does is that it essentially evaluates contractions over masked matrices in a rather unconventional way by using while loops to do the contraction over the two matrix indices (which are used to avoid recompilation if the lengths of the indices allowed by the masks change, which will be the case in our application). The defined function takes the matrix which should be contracted, batches of valid indices for the first dimension (padded with "-1"s), a single array of valid indices for the second matrix dimension (again padded with "-1"s) and returns the batch of contracted matrix values. Interestingly, when the contracted value is chosen to be \sum matrix[i,i] * j rather than the value we actually want, \sum matrix[i,j], the code runs a lot faster (approx. factor 10 for the chosen dim=500 example). This can definitely not be right or am I missing something? Unfortunately I have not been able to track this issue further down the stack, the only other interesting observation I made is that the timing discrepancies seem to go away when the vmap across the batch dimension is left out (which to me looks like an indication that there is some issue with the vmapping).
The minimal example is defined as follows:

```python
import jax
import jax.numpy as jnp

def masked_contraction(my_array, stacked_mask_indices_i, mask_indices_j, comparison=False):
    def contract_array(mask_indices_i):
        """The loop_state arguments are just tuples where the first item holds the
        count and the second the accumulated value.
        """
        def first_index_loop(outer_loop_state):
            i = mask_indices_i[outer_loop_state[0]]
            def compute_inner(_):
                def inner_loop(inner_loop_state):
                    j = mask_indices_j[inner_loop_state[0]]
                    if comparison:
                        # This case is much faster than the one below
                        return (inner_loop_state[0] + 1, inner_loop_state[1] + my_array[i, i] * j)
                    else:
                        # This is the case we are actually interested in (which is very slow)
                        return (inner_loop_state[0] + 1, inner_loop_state[1] + my_array[i, j])
                return jax.lax.while_loop(lambda x: mask_indices_j[x[0]] != -1, inner_loop, (0, 0.))[1]
            """
            The following cond is pretty much useless in this case but it seems to amplify the issues,
            without it the runtime differences are less substantial (still significant though).
            """
            value = jax.lax.cond(i<my_array.shape[0], compute_inner, lambda _: 0., None)
            return (outer_loop_state[0] + 1, outer_loop_state[1] + value)
        return jax.lax.while_loop(lambda x: mask_indices_i[x[0]] != -1, first_index_loop, (0, 0.))[1]
    # Without the following vmap the timing issues do not appear to be a problem
    return jax.vmap(contract_array)(stacked_mask_indices_i)
```

The comparison flag is really just for testing, the case we are interested in is the (much slower) one with comparison=False. I tested the code with python 3.9 and the most recent jax version available via pypi (jax: 0.2.27, jaxlib:0.1.75) with the following test in ipython:
```python
masked_contraction_jitted = jax.jit(masked_contraction, static_argnums=3)

testarray = jax.random.uniform(jax.random.PRNGKey(1), (500,500))

first_index_mask = (jnp.arange(500) % 2 == 0)
second_index_mask = (jnp.arange(500) % 3 == 0)

mask_indices_i = jnp.nonzero(first_index_mask, size=500, fill_value=-1)[0]
mask_indices_j = jnp.nonzero(second_index_mask, size=500, fill_value=-1)[0]

mask_indices_i_stacked = jnp.stack((mask_indices_i, mask_indices_i))

# run once before timing to ignore any compilation overhead
masked_contraction_jitted(testarray, mask_indices_i_stacked, mask_indices_j, True)
%timeit masked_contraction_jitted(testarray, mask_indices_i_stacked, mask_indices_j, True)

# run once before timing to ignore any compilation overhead
masked_contraction_jitted(testarray, mask_indices_i_stacked, mask_indices_j, False)
%timeit masked_contraction_jitted(testarray, mask_indices_i_stacked, mask_indices_j, False)
```

and got the following timings:

> %timeit masked_contraction_jitted(testarray, mask_indices_i_stacked, mask_indices_j, True)
> 491 µs ± 10.2 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)
> %timeit masked_contraction_jitted(testarray, mask_indices_i_stacked, mask_indices_j, False)
> 62.9 ms ± 27.9 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)

It would be great if somebody could look into this and help me understand what is going on. Please let me know if you require further information or if I should post this somewhere else (at the moment this really seems like a bug to me so I'm posting it here).

Thank you very much and best wishes,
Yannic

**Comment by patrick-kidger** [comment]
At first glance it looks like this might be the vmap-of-cond-is-select bug/limitation. Namely, that an unvmap'd `lax.cond` only evaluates the branch it needs to, but a vmap'd `lax.cond` turns into a `lax.select`, and evaluates both branches unconditionally (before merging the results together).

**Comment by yannra** [comment]
Hi @patrick-kidger,
Thanks for the reply, that's an interesting shout! Could you point me to the bug description for that issue, I'd need to read up a little bit more on that to fully understand this issue. However, I don't really see why this bug would give us two different runtimes for the two different cases tested (if we use array[i,i] *j instead of array[i.j] inside the inner loop the computation is much faster). If really both branches are evaluated when the vmap is applied, this should be the same in both test cases (and thus slowing down both calculations) or not?

**Comment by jakevdp** [comment]
Just looking at the code, I suspect the performance issue comes from the compiler's ability to parallelize the expensive parts of your computation. Your inner loop has two gather (i.e. indexing) operations: in the fast version, they can be done in parallel. In the slow version, the output of the first (`j`) becomes the input of the second, so they must be done sequentially.

You could dig into this more using some of JAX's profiling tools: https://jax.readthedocs.io/en/latest/profiling.html

**Comment by yannra** [comment]
Thanks for taking a look at the code! It must be something like that, however the slowdown goes away when not applying the vmap (i.e. when I just call the function for the two different masks one after the other the timings are as expected even with the full indexing inside the innermost loop), so I really suspect there is some compiler optimization which is not correctly applied in the vmaped case. I'll try to get some profiling data later and post them here (unfortunately I don't really understand the generated XLA language enough to chase down this bug on the XLA level)...

**Comment by yannra** [comment]
Sorry for leaving this open for so long. I think I've got a much better understanding about what is going here and will close this issue.
I believe the performance discrepancies really just come from what has been stated before: Without the vmap, the cond (and I presume also the while loop) shortcuts so that only the true cases are evaluated (on the CPU) which is no longer possible with the vmap. I'm not 100% sure how while loops nested inside a vmap behave but I suspect that these are also extended to have the same "loop length" over the full batch dimension. The comparison case in the code above was really not well chosen as the indexing can be parallelised.
Anyway, thanks for your input, allowing me to understand jax a little bit more!  We have found better implementations avoiding these limitations.
I'm closing this issue just to have no more

**Comment by yannra** [comment]
Sorry for leaving this issue untouched for so long. I think I have a much better understanding of what is going on now and I see where the performance discrepancies come from. I'm therefore finally closing this issue (after I had forgotten about it a little bit to be honest).
It should really be what has been stated before: Without the vmap, the cond (and likely also the while) can shortcut so that the executed loop is constrained to the true cases, which is no longer possible with the vmapping. I don't fully understand how while loops inside a vmap behave but it must be similar (in the sense that all while loops are to some degree extended to the same execution length for each individual run inside the vmap). The comparison case above with the alternative indexing was indeed not really sensible as this allows for much more compiler optimization + parallelization (explaining why  this was faster).
Anyway, thanks again for the really helpful input, which definitely helped me understand jax a little bit better!
